policy/daemon.go

```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
 		glg.Warn(errors.Wrap(err, "error body.close"))
 	}
 
-	// set eTag cache
-	eTag := res.Header.Get("ETag")
-	if eTag != "" {
-		glg.Debugf("Setting ETag %v for domain %s", eTag, domain)
-		p.etagCache.SetWithExpire(domain, &etagCache{eTag, sp}, p.etagExpTime)
+	// verify policy expiry
+	var etagValidDur time.Duration
+	if sp.SignedPolicyData.Expires == nil {
+		glg.Warnf("Warning: policy expiry not set, domain: %v", domain)
+		etagValidDur = p.etagExpTime - p.expireMargin
+	} else {
+		// if not set, sp.SignedPolicyData.Expires is nil
+		// if set, but invalid, sp.SignedPolicyData.Expires is Time{}
+		validDur := sp.SignedPolicyData.Expires.Time.Sub(fastime.Now())
+		if validDur <= p.expireMargin {
+			p.etagCache.Delete(domain)
+			glg.Errorf("Error verifying policy expiry, policy expiry: %v, valid duration: %v, valid margin: %v", sp.SignedPolicyData.Expires.Time, validDur, p.expireMargin)
+			return nil, false, errors.New("policy already expired")
```

Comment:
this logic is wrong, the policy is not expired yet but this error returned.

Reply:
It depends on the user's config. If the refreshDuration > expireMargin, there will be an empty policy time gap.

Either keep this, or add the `refreshDuration > expireMargin` checking in the constructor to warn the user?

Reply:
no, it depends on the policy expire time, 
for example policy expire time is 1 week, refresh duration is 5 days and expire margin is 1 hour, it should work well. 
expire margin means that the margin time to force update the policy without using e-tag.

Reply:
for your case, if the policy expire time is 1 week and user set the margin to 1 month then the error will always be returned.

Reply:
move to sp.verify()